### ISSUE TITLE
Replace Calc methods with Vecmath, instead of JAMA

### ISSUE THREAD

**Issue by lafita** [biojava/biojava#303]
The Calc class contains methods to calculate geometric operations on the Structure interface objects (basically Atoms). 
Now that the Vecmath dependency has been added to the structure module, some methods have been added to convert from the old JAMA matrices and vectors to the Vecmath objects. 

Some problems have already occurred in the conversion, basically because the JAMA rotation matrices are pre-multiplication, and Vecmath uses post-multiplication matrices to transform points and vectors (which is the convention in 3D geometry). See 38c1e33. 
This means that to obtain the same transformation the matrices have to be transposed (inverted). The current Calc methods that convert between the two matrix representations ensure that this conversion is made (since 38c1e33).

However, it would be a step forward to replace completely all JAMA matrices in the Calc class to the Vecmath ones, so that no conversion methods or different matrix conventions are used. 
This would require work ensuring that the basic algorithms, like SVD calculation, return the appropriate transformation matrix in the post-multiplication form (as Vecmath requires).

Another TODO would be to create some tests for the Calc class (if possible, before changing to the vecmath implementation of the algorithms) that ensure the same behavior as the old code.

**Comment by andreasprlic** [comment]
To be more backwards compatible, it would be great if Calc would initially support both Vecmath and JAMA...

**Comment by lafita** [comment]
Both are supported right now in Calc, but there is this confusion with the transposition of the rotation Matrix. The bug should be fixed now.
The only thing to be done is a test, then?

**Comment by josemduarte** [comment]
From my understanding the key piece of code is in `Calc.rotate(Atom,Matrix)` and `Calc.rotate(Atom,double[][])`. There the vector is pre-multiplied. Since the usual convention is to post-multiply, this means that all rotation matrices that use `Calc.rotate` are in effect transposed with respect to the usual rotation matrices found in textbooks or elsewhere. 

That really can cause some confusion and should be avoided. I'd definitely advocate for unifying the code as much as possible, in my opinion the best solution would be moving everything to vecmath, making sure that there's nothing needed from JAMA (SVD perhaps?). As @lafita says, provided we have good tests for it, doing this change should not be too hard. Surely the hardest part is to write good tests that cover well the different cases.

**Comment by andreasprlic** [comment]
Agreed, removing all JAMA usage would be the clean thing to do. We have a few unit test that should help with making sure the results are numerically the same. Only issue is that it is not the most exciting thing to do. Once we have the code for faster implementation of RMSD and translations from Peter, we can even deprecate SVD.

**Comment by lafita** [comment]
If we make sure that all methods (like SVD) generate post-multiplication rotation matrices (doesn't matter if they are Vecmath or JAMA), then we can keep the support for JAMA in Calc. The only thing to be changed, which is the confusing one, would be the behavior of the methods:
<pre>Calc.rotate(Atom,Matrix)
Calc.rotate(Atom,double[][]) </pre>
That way we could keep the backwards compatibility with JAMA in Calc, although some users might have problems with older code that use the past versions of those two methods.

**Comment by josemduarte** [comment]
It seems that vecmath does have SVD after all, see the [Gmatrix class](http://docs.oracle.com/cd/E17802_01/j2se/javase/technologies/desktop/java3d/forDevelopers/j3dapi/javax/vecmath/GMatrix.html).

**Comment by sbliven** [comment]
I think changing the convention of an existing method is a very bad idea. Instead, lets just deprecate the Matrix and double[][] versions and use the Matrix3d/4d versions going forward.

**Comment by andreasprlic** [comment]
for now let's deprecate all methods that use JAMA

**Comment by lafita** [comment]
I have been reviewing the code that uses **JAMA** matrices. This is a summary:

1. Structural alignment data structures (`AFPChain`, `FragmentPair`, etc) store the transformation as a rotation matrix (**JAMA**) and a translation vector (`Atom`).
2. Calc uses rotation matrices (**JAMA**) to rotate `Atom `coordinates, but has equivalent methods to do the same with **vecmath** matrices.
3. **DP** similarity matrices in the structural alignment calculators (like `CECalculator`) stored as `double[][]` get transformed sometimes to **JAMA** Matrices, with no apparent reason (the double array could be used instead).
4. `SuperPositionSVD` uses the SVD decomposition on a **JAMA** matrix. 

I propose the following solutions, which would completely remove the **JAMA** dependencies in **BioJava**: 

1. Store the transformations as a single 4x4 `Matrix4d` object, which replaces the rotation matrix and translation vector. This simplifies inverting of the transformation and maintaining the information consistent, since there is a single object. The rotation matrix and translation vector can always be extracted from the `Matrix4d `transformation (maybe some utility methods can be provided for that).
2. Keep only the methods in `Calc `that operate with **vecmath** objects, and remove the ones that use **JAMA** matrices. 
3. Never convert to **JAMA** matrices the internal `double[][]` matrices. Critical is to make sure column and row indices are not swapped in doing so. If needed, use `GMatrix `objects.
4. As @josemduarte indicated, there is an equivalent method to do SVD in **vecmath**, using `GMatrix`.

**Comment by lafita** [comment]
I am currently midway in the refactoring. I have realized that **vecmath** does support `SVD`, but the features and **API** are very poor (only the three matrices are returned), while **JAMA** has a nice way to acces the array of single values directly. In addition, **JAMA** also supports eigenvalue decompositions with a nice interface, instead of calculating them from the **SVD**.

I think it makes sense to keep **JAMA** for these types of linear algebra calculations with big matrices, while replacing all the geometry code with **vecmath** (much better and no confusions). Step **4** will not be changed then. 

What do you think?

**Comment by sbliven** [comment]
Let's try to get the public API switched to vecmath for biojava 5. Internal uses of JAMA (eg for SVD) can be removed later without breaking compatibility.

**Comment by josemduarte** [comment]
@lafita  you mentioned that you were working in this refactoring at some point, do you still have that code around in a branch somewhere? It'd be good to include some clean up related to this in the 5 release.

**Comment by lafita** [comment]
We already improved a lot the documentation of the `Calc` methods so that confusion is reduced and we introduced the `CalcPoint` class that uses only vecmath objects (no Atoms or JAMA matrices). See #548 and #583.
Also, the new Superposition interface makes all methods equivalent no matter what internal data structures they use. See #579. 
I think this already solves many of the issues that we were encountering.

I gave up in trying to replace JAMA for two reasons:
1. JAMA matrices are used internally in many algorithms with complicated nested loops that are hard to debug. 
2. JAMA supports eigenvalue decompositions with a nice interface. In vecmath they have to be extracted from the SVD, which requires writing additional methods prone to bugs.
